### ISSUE TITLE
Savestate support - not working (Switch, Android)

### ISSUE THREAD

**Issue by tomjramsden** [libretro/libretro-uae#219]
Hello,

I saw the savestate is feature is now supported!! 😀😀
I’m on RA 1.8.3 (latest build) as of 13/01.. on Nintendo Switch. I presume switch is supported?

When I tried testing it yesterday it crashed retroarch
Can anyone replicate? I also haven’t tried on my pc yet...

@rsn8887 - can you replicate this error on the Switch also?

**Comment by MajorWilly** [comment]
Confirmed. When making a savestate (with a WHDLoad LHA game) it crashes. Floppy savestates work fine.

Tested on Android.

**Comment by tomjramsden** [comment]
Thanks for confirming @MajorWilly 

> It's probably better to make an official save to WHDSaves.hdf for games that support it anyway). Floppy savestates work fine.

How does whdsaves.hdf work? Is it for in game saves?

**Comment by rsn8887** [comment]
I don't think hdf savestates have ever worked. Not even in WinUAE.

**Comment by tomjramsden** [comment]
> Does it crash with .adf also or only .hdf? I don't think hdf savestates have ever been supported by any UAE version. Not even in WinUAE.

I have only tested with .lha I will test a .hdf later

**Comment by rsn8887** [comment]
.lha is still using hdf internally. As I said savestates with hdf are not really supported on any platform.

Please try .adf and if that works, then the feature works.

**Comment by GuybrushThreepwood-GitHub** [comment]
I've tested Switch and Android savestates with hdf and lha and neither work, they crash RetroArch.

Curiously the latest Win64 RA and the latest builds of the emu, they seem to save the hdf/lha state fine (tested Addams Family) - haven't gone too deep perhaps eventually with multiple save/loads things stop working but it does restore fine from my limited tests.

**Comment by tomjramsden** [comment]
Yeah according to this the latest release notes and this comment https://github.com/libretro/libretro-uae/pull/123#issuecomment-532315728 it should be working with .hdf and that is somewhat confirmed from the comment above. I wonder if it has to do with the way the filesystem is accessed on switch?

@sonninnos can you please confirm the expected behaviour of the new feature? especially around supported formats and platforms (appears to be working in Windows)

**Comment by sonninnos** [comment]
The problem with HDFs and savestates come only when there is reads and writes to HDF after said savestate load. So with mostly preloaded WHDLoads there is no problem.

Could you try an older build if it crashed always? I don't get crashes with Windows. but maybe that recently added `savestate_initsave()` function does it. I'll remove it, since there is no real benefit after all.

**Comment by tomjramsden** [comment]
Thanks for the update, verified Windows works fine, but switch is stillbroken. I have tried latest and a couple of older versions but it doesn’t work, apparently android doesn’t either.

**Comment by rsn8887** [comment]
What about Windows 64 bit, does it work there? (Wondering if this could be a 32bit/64bit issue, since both Android and Switch builds are 64bit AFAIK.)

**Comment by sonninnos** [comment]
> Also, whilst your here could please take a look at #215

Sorry, nothing I can do about that, I'm on Windows 10 64bit and can't replicate.

**Comment by rsn8887** [comment]
You are on Windows 10 64bit and also using x64 bit build of Retroarch?

I just tried to save state on MacOS, and it crashed on saving. MacOS is also 64 bit.

**Comment by sonninnos** [comment]
Yes, 64-bit version of RA too.

**Comment by rsn8887** [comment]
On MacOS it is indeed fixed.

**Comment by MajorWilly** [comment]
Windows (64-bit) works, Android (32-bit) still crashes when saving state of a WHDLoad LHA (and HDF).

**Comment by tomjramsden** [comment]
Confirmed that is still unfortunately this is still broken for switch using Retroarch 1.8.4 and latest build.

**Comment by tomjramsden** [comment]
> Windows (64-bit) works, Android (32-bit) still crashes when saving state of a WHDLoad LHA (and HDF).

Is this still broken for you? Did you say adf save states are working on android?

**Comment by MajorWilly** [comment]
Yeah, ADF savestates work fine on Android. Well, I only have 32-bit Android, I don't know if it works on 64-bit. WHDLoad LHA and HDF's crash as soon as you select 'savestate'.

**Comment by sonninnos** [comment]
So CD32 savestates should be OK too?

It can't be too complicated to find the reason for that HDF issue, but I guess it could also be circumvented with LHA cases by not using WHDLoad.hdf as HDF but instead put the files raw inside a directory. That should also remove the ill effects of HDF savestates not working properly in general, since then there would be no HDFs that get written to (of course WHDSaves.hdf should also be changed to a directory, as is already possible.)

Of course fixing the crash is more important, but I have no way of doing anything about it.

**Comment by MajorWilly** [comment]
I haven't got round to checking CD32 save states yet, I deleted them off my device to get some space back. The weird thing is that save states with LHA files work fine on Windows (and someone said Mac too) and as all saves are redirected to WHDSaves then HDF's should be untouched ? Especially if they're Zipped because Retroarch will unzip a fresh copy on every run.

**Comment by sonninnos** [comment]
Not really weird since the crash is (most likely) not related to UAE not being 100% compatible with hard drives and savestates. As in fixing the other won't fix the other, and by my understanding that 100% compatibility will never happen, so fixing the crash is the best possible outcome.

Yes in theory there should not be any more writes to WHDLoad.hdf after the initial Kickstart copying, but the `WHDLoad.prefs` core option will copy `WHDLoad.prefs` from RA system to the WHDLoad.hdf on every boot, no matter the setting. Without it there still is no guarantee that in the future the hard drive is in the same state as when the particular savestate was saved. All possible tricks (AFAIK) are already done to minimize the writes to the HDF (RAM temp use etc.), and those won't help with non-WHDLoad HDFs.

Nothing should get written to game HDFs anymore and certainly nothing is written to read only LHAs. But it won't matter (with the crash issue and/or with the compatibility issue) if there is even one HDF (WHDload.hdf) attached.

**Comment by MajorWilly** [comment]
@sonninnos Just out of interest, what was the theory on why savingstate with a LHA works on Windows but not on Android/Switch ?

I'm a bit confused about the issue title ?

@tomjramsden Do ADF savestates also crash on Switch or is it just with WHDLoad LHA/HDF like on Android ?

@rsn8887 said it was fixed on MacOS. WHDLoad LHA savestates work ?

**Comment by sonninnos** [comment]
I haven't seen any theory yet.

**Comment by GuybrushThreepwood-GitHub** [comment]
Ok done some more testing today...

* Win10 (64 bit RA) - HDF, LHA and ADF all save/restore the state fine.
* Android (64 bit RA) - HDF and LHA crash RA completely but ADF savestates are good.
* Switch - HDF/LHA do not work, ADF does work but you cannot cycle through the savestate slots to make different saves, eventually it either crashes, hangs or it sometimes starts saying 'core does not support savestates' :) - plus on switch you eventually run into issue #215 which requires a complete reboot.

Have to say regardless of the savestate issues I appreciate all the work done on the emulator. Top job.

**Comment by mors-gotha** [comment]
[...]
> * Switch - HDF/LHA do not work (it will crash RA), ADF does work but you cannot cycle through the savestate slots to make different saves, eventually it either crashes, hangs or it sometimes starts saying 'core does not support savestates' :) - plus on switch you eventually run into issue #215 which requires a complete reboot.
[...]

Same thing with PUAE in retroArch using linux (centos 7).
When i try to save a savestate with a LHA game from retroplay, RA crashes and the savestate doesn't work. With ADF/IPF games it works.

**Comment by sonninnos** [comment]
Thanks for the reports. Let's hope this change will work on all platforms.

**Comment by mors-gotha** [comment]
Unfortunately it still doesn't work in linux. Retroarch crashes when trying to save a savestate with a LHA game and the savestate doesn't load when you reload retroarch (it crashes again).

**Comment by sonninnos** [comment]
The fix was found and confirmed with linux, so are you sure buildbot has compiled the latest version yet?

**Comment by mors-gotha** [comment]
Hmm, maybe you're right. I saw a new build created today at 14:18 and thought it already included the fix. I apologize if I made a mistake!

**Comment by mors-gotha** [comment]
Ok, i confirm the last build fixed the problem! Now you can use savestates with LHA games!
Thank you!

**Comment by tomjramsden** [comment]
> Thanks for the reports. Let's hope this change will work on all platforms.

Will try to test on switch later on, I was having issues last night where the core was crashing on every game, but I think it's to do with the online updater again. Fingers crossed!

**Comment by GuybrushThreepwood-GitHub** [comment]
I can confirm Android now saves HDF/LHA without issue. Great work.

I'll try and test Switch when I can get a moment.

**Comment by tomjramsden** [comment]
> I can confirm Android now saves HDF/LHA without issue. Great work.
> 
> I'll try and test Switch when I can get a moment.

The whole core is broken on Switch for me. It crashes when I attempt to load any content from a playlist or through load content? Can you replicate please?

**Comment by rsn8887** [comment]
Can you change option Use WHDSaves to True and try again?

**Comment by sonninnos** [comment]
Whatever it is, it is not savestate related, so no reason to reopen this issue.

Also try deleting the core option config and start with a fresh one. And any kind of log is better than none.

**Comment by tomjramsden** [comment]
> "Whole core broken" is not really a useful bug report. - agreed. I’ve changed it to a more accurate comment.
> 
> What about starting core without content? Loading core works without crash, however I can’t the start the core. I get a ‘failed to load content error’
So, as a workaround I went in and check the core-options config file and checked WHDSaves is already set to true.
> What about .adf loading, does it work? - I’ll try that later.
> Can you start core (without loading content), change option `Use WHDSaves` to `True` and try again? See above

**Comment by sonninnos** [comment]
Something wrong with the Kickstarts then if you can't start the core at all. Please find out how to get the stdout verbose log.

**Comment by GuybrushThreepwood-GitHub** [comment]
Ok had a look just now (build 1da3fa0) I'll try and explain this but I think issue #215 is part of the real problem now.
* HDF/LHA savestates do save now and *up to a point* I can reload/save and cycle the state number and apart from a small graphical glitch and pause when loading - it does work.
* ADF is roughly the same, it works initially, however it seems to die quicker after about 2-3 saves/restores.

At the moment you cannot run more than one game after another - closing content and/or closing RA basically causes the entire Switch to crash when you try to load RA or another game and you have to reboot (I'm running SX OS 2.9.3 if this matters).

After a reboot if I had an HDF/LHA savestate of a game it does load 'sometimes' (probably just luck) but states that did work on the initial run and save, can/will die when running the game again.

**Comment by tomjramsden** [comment]
> Something wrong with the Kickstarts then if you can't start the core at all. No reason to bother testing ADFs until you get the Kickstart screen with just starting the core. Please find out how to get the stdout verbose log.

I reloaded the kicks over ftp and removed the core-options, it is now working again! Thanks

**Comment by tomjramsden** [comment]
> Ok had a look just now on Switch (build [1da3fa0](https://github.com/libretro/libretro-uae/commit/1da3fa0b82acf1d77d302b3f055d75cb02299f97)) I'll try and explain this but I think issue #215 is part of the real problem now.
> 
> * HDF/LHA savestates do save now and _up to a point_ I can reload/save and cycle the state number and apart from a small graphical glitch and pause when loading - it does work.
> * ADF is roughly the same, it works initially, however it seems to die quicker after about 2-3 saves/restores.
> 
> At the moment you cannot run more than one game after another - closing content and/or closing RA basically causes the entire Switch to crash when you try to load RA or another game and you have to reboot (I'm running SX OS 2.9.3 if this matters).
> 
> After a reboot if I had an HDF/LHA savestate of a game it does load 'sometimes' (probably just luck) but states that did work on the initial run and save, can/will die when running the game again.
> 
> Actually as a side note - this is very similar issue that FS-UAE has with the WHDLoad savestates, they can work initially but on a reload in another session, they stop working.

Yes this is still very much an issue for me also. Interestingly I am using the same version of SXOS. Perhaps some more testing is needed on #215 out of interest are using a nsp redirector for Retroarch?

**Comment by sonninnos** [comment]
> Actually as a side note - this is very similar issue that FS-UAE has with the WHDLoad savestates, they can work initially but on a reload in another session, they stop working.

Similar to any UAE like I previously tried to explain. Hardfiles and savestates will never be perfect together, according to the authors. And If you try to use in-game saves after loading savestates you will definitely see a lot of guru meditations, crashes and weird behavior.

Here savestates work fine on later sessions, but a while back the savestates in general (yes, even with ADFs) had a problem which I thought I fixed, but apparently not (has not happened here in a while though). The best and only way to fight around it is to always make sure that loading works after saving, because the save could for some unknown reason get corrupted. The savestate undo features in RA are very helpful in this case.

Never happened like that on WinUAE though, so it could be either this old version of UAE or the way this core calls the saving. I'll sometime have to go through all the savestate parts of the code to see if there are differences.

When testing CD32 games so far I've found only one game that has random issues of not always loading the next level after loading states, and that is Alien Breed Tower Assault.

And Brutal Football will start playing CD audio after loading a state saved in the menu, where CD audio does not even play when saved, but that is a minor issue. I'll have to test if it happens with WinUAE too.

**Comment by GuybrushThreepwood-GitHub** [comment]
Ah ok - apologies I skimmed over the comments and missed the explanation. Even in FS-UAE they do specify the config as 'unsafe_save_states' I wasn't sure how far removed the various UAE codebases might be.
As I said, I really appreciate what you've done here - nice to have the emu on the go and savestates is just a cherry on top even with some limits.

> out of interest are using a nsp redirector for Retroarch?

I use the plain old zip extract for each build, this update test was the first time I actually used the online updater on Switch to get the latest nro as I prefer to not possibly get a bad update and have to rebuild anything on the SD card again.

Actually I assume now the Linux version has been fixed, the various Switch Lakka/Ubuntu builds of RA would actually be an option for running the emu with a little more stability.
